cmd/kubeless/function/top.go

```diff
@@ -0,0 +1,270 @@
+type PrometheusMetricsHandler struct{}
+
+func getFunctions(kubelessClient versioned.Interface, namespace, functionName string) ([]*kubelessApi.Function, error) {
+	var functionList *kubelessApi.FunctionList
+	var f *kubelessApi.Function
+	var err error
+	if functionName == "" {
+		functionList, err = kubelessClient.KubelessV1beta1().Functions(namespace).List(metav1.ListOptions{})
+		if err != nil {
+			return []*kubelessApi.Function{}, err
+		}
+	} else {
+		f, err = kubelessClient.KubelessV1beta1().Functions(namespace).Get(functionName, metav1.GetOptions{})
+		if err != nil {
+			return []*kubelessApi.Function{}, err
+		}
+		functionList = &kubelessApi.FunctionList{
+			Items: []*kubelessApi.Function{
+				f,
+			},
+		}
+	}
+	return functionList.Items, nil
+}
+
+func parseMetrics(namespace, functionName string, rawMetrics []byte) ([]*Metric, error) {
+	parser := expfmt.TextParser{}
+	parsedData, err := parser.TextToMetricFamilies(bytes.NewReader(rawMetrics))
+	if err != nil {
+		return nil, err
+	}
+
+	tmp := map[string]*Metric{}
+	var parsedMetrics []*Metric
+
+	metricsOfInterest := []string{"function_duration_seconds", "function_calls_total", "function_failures_total"}
+	for _, m := range metricsOfInterest {
+		for _, metric := range parsedData[m].GetMetric() {
+			// a function can have metrics for multiple methods (GET, POST, etc.)
+			// method names can be values other than GET/POST/PUT/DELETE
+			for _, label := range metric.GetLabel() {
+				if label.GetName() == "method" {
+					if _, ok := tmp[label.GetValue()]; !ok {
+						tmp[label.GetValue()] = &Metric{
+							FunctionName: functionName,
+							Namespace:    namespace,
+							Method:       label.GetValue(),
+						}
+					}
+					if m == "function_failures_total" {
+						tmp[label.GetValue()].TotalFailures = metric.GetCounter().GetValue()
+					}
+					if m == "function_duration_seconds" {
+						tmp[label.GetValue()].TotalDurationSeconds = metric.GetHistogram().GetSampleSum()
+					}
+					if m == "function_calls_total" {
+						tmp[label.GetValue()].TotalCalls = metric.GetCounter().GetValue()
+						if tmp[label.GetValue()].TotalCalls > 0 {
+							tmp[label.GetValue()].AvgDurationSeconds = float64(tmp[label.GetValue()].TotalDurationSeconds) / tmp[label.GetValue()].TotalCalls
+						}
+					}
+				}
+			}
+		}
+	}
+
+	// if the funciton hasn't been invoked, add an item to the list so the function displays in the output
+	if len(tmp) == 0 {
+		tmp[""] = &Metric{
+			FunctionName: functionName,
+			Namespace:    namespace,
+		}
+	}
+
+	for _, v := range tmp {
+		parsedMetrics = append(parsedMetrics, v)
+	}
+
+	return parsedMetrics, nil
+}
+
+func (h *PrometheusMetricsHandler) getRawMetrics(apiV1Client kubernetes.Interface, namespace, functionName string) ([]byte, error) {
+	svc, err := apiV1Client.CoreV1().Services(namespace).Get(functionName, metav1.GetOptions{})
+	if err != nil {
+		return []byte{}, err
+	}
+
+	port := strconv.Itoa(int(svc.Spec.Ports[0].Port))
```

Comment:
the logic to get the port of a function is already in `pkg/utils/event_sender.go` can you refactor it so these two functions use it?